### PR TITLE
nixos/flatpak: pass system icons and fonts

### LINKED ISSUE #119433: Flatpak can't access system fonts and icons
**Describe the bug**
Flatpak mounts /usr/share/icons and /usr/share/fonts to /run/host/share/icons and /run/host/fonts respectively. However, this doesn't work on NixOS since all fonts/icons are in their own directories in /nix/store. While the lack of icon theme in flatpak apps is not so a big problem, the lack of system fonts leads to inability to read text written in some languages. The font for Latin characters is Deja Vu is that case that is too bold and has bad readability (at least for me). I'm suffering from this issue for a long time and thought that at some point someone raise this or this will be fixed at some time by someone, but I can't take it anymore :(

**To Reproduce**
1. Install any app from flatpak
2. Realize that it uses Deja Vu font and can't display various language families (e.g. CJK), can't display graphical symbols, etc.
3. Realize that it can't use system icons even if it tries

**Expected behavior**
Flatpak is patched so that it can read fonts and icons. Or some option that enables centralized storage of fonts/icons and Flatpak uses it.

**Screenshots**
изображение
изображение

How it looks like outside of flatpak:
изображение

**Notify maintainers**
@jtojnar

**Metadata**
```
 - system: `"x86_64-linux"`
 - host os: `Linux 5.11.11-zen1, NixOS, 21.05pre-git (Okapi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.10`
 - nixpkgs: `/run/nixpkgs`
```

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute: flatpak
# a list of nixos modules affected by the problem
module: services.flatpak
```

### PR DESCRIPTION
**Pull request by io12**
Fixes https://github.com/NixOS/nixpkgs/issues/119433

## Description of changes

Add another patch to Flatpak that mounts the NixOS system font and icon directories into the sandbox if the standard FHS ones don't exist. The FHS paths are still checked so that the Flatpak build still works on non-NixOS systems. Since these directories contain symlinks into `/nix/store`, the patch also walks the tree of symlinks to collect a list of `/nix/store` paths to expose to the sandbox.

Also, change the default for `fonts.fontDir.enable` from `false` to `services.flatpak.enable`.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/libraries/flatpak/fix-fonts-icons.patch
Comment (jtojnar) [inline pkgs/development/libraries/flatpak/fix-fonts-icons.patch:28]: This should be freed. You can have it done [automatically](https://github.com/GNOME/glib/blob/4144341e7a8a86c5eb980a0093e35c7606b9bff5/glib/glib-autocleanups.h#L55) with:

```suggestion
+      g_autoptr(GDir) dir = g_dir_open(source_path, 0, NULL);
```

File: pkgs/development/libraries/flatpak/fix-fonts-icons.patch
Comment (jtojnar) [inline pkgs/development/libraries/flatpak/fix-fonts-icons.patch:34]: Similarly, this can be simplified, also you can use [`g_build_filename`](https://docs.gtk.org/glib/func.build_filename.html):

```suggestion
+          g_autofree gchar *path = g_build_filename (source_path, file_name, NULL);
+          get_nix_closure (closure, path);
```

File: pkgs/development/libraries/flatpak/fix-fonts-icons.patch
Comment (jtojnar) [inline pkgs/development/libraries/flatpak/fix-fonts-icons.patch:24]: Another place where automatic memory management can simplify the code a bit, combined with [`g_steal_pointer`](https://docs.gtk.org/glib/func.steal_pointer.html):

```suggestion
+      g_autofree gchar *path = g_malloc(PATH_MAX);
+      realpath(source_path, path);
+      if (g_str_has_prefix(path, "/nix/store/"))
+        {
+          *strchr(path + strlen("/nix/store/"), '/') = 0;
+          g_hash_table_add(closure, g_steal_pointer (&path));
+        }
```

File: nixos/modules/config/fonts/fontdir.nix
Comment (jtojnar) [inline nixos/modules/config/fonts/fontdir.nix:32]: Maybe enable it explicitly in the flatpak module. That way people will not be able to just override it accidentally. It will also be properly shown in nixos-option as defined by flatpak.

**Review by jtojnar — COMMENTED** [review]
This will not handle symlinks pointing to trees of symlinks (e.g. when using font collections produced with `buildEnv`) but I am not sure how common that is.

I also wonder about performance with large font libraries. I thought about using `nix-store --query --references` but that will not work for the `/run/current-system` paths directly, since they are not in Nix store. It would also pull in all dependencies, not just those referenced by icons/fonts. But I guess we can just address it by having NixOS generate `/run/current-system/sw/share/nix-flatpak/fonts.closure` if/when someone notices a performance degradataion.

**Comment by io12** [comment]
Thanks for the review! I applied the suggestions.

> This will not handle symlinks pointing to trees of symlinks (e.g. when using font collections produced with buildEnv) but I am not sure how common that is.

I think it does, because `get_nix_closure()` is recursive. I think I remember this being a requirement for icons.

**Comment by andrevmatos** [comment]
I'm running this locally, it seems to work fine, at least for fonts (I can see my system fonts in Inkscape without the bindfs workaround).
cursors though seems to work in some apps, and fallback to a black cursor in Gtk3+ ones, not sure if something with this patch or my local config (KDE, cursors set in Plasma and Gtk configs to Oxygen White)

**Comment by io12** [comment]
> I'm running this locally, it seems to work fine, at least for fonts (I can see my system fonts in Inkscape without the bindfs workaround). cursors though seems to work in some apps, and fallback to a black cursor in Gtk3+ ones, not sure if something with this patch or my local config (KDE, cursors set in Plasma and Gtk configs to Oxygen White)

Does the bindfs workaround have the same behavior? If no, what is the output of `flatpak --user override --show`?

**Review by andrevmatos — APPROVED** [review]
After some restarts (maybe cleaning up the old /usr bindfs mounts), and checking again, this patch seems to be working fine. I can see fonts everywhere, and proper cursors in most apps (but a few stubborn gtk ones, but would not make an issue for them). LGTM

**Comment by AndreiSva** [comment]
will this fix the mouse cursor not being the adwaita theme in flatpak apps?

**Comment by io12** [comment]
> will this fix the mouse cursor not being the adwaita theme in flatpak apps?

Yes

**Comment by beh-10257** [comment]
since well this is still an open issue is passing themes as well out of the question

**Comment by io12** [comment]
> since well this is still an open issue is passing themes as well out of the question

I don't think flatpak tries to pass themes by default, just fonts and icons. I also don't know if nixpkgs would be willing to accept a patch to add a feature that doesn't exist on other distros like that?

**Comment by beh-10257** [comment]
This is what I used to do on arch symlink th theme I want to use from /usr/share/.../themes
To .local/share/themes
And yeah that's enough obv after allowing flatpaks to read
.local/share/themes directory

it seems just installing the package isn't enough in nixos or home-manager since I have no idea where are the themes even stored
Also since well I don't mean make it by default but having the option would just be nice

Also so adding icon packs to .local/share/icons manually isn't gonna pass them ??
Even with this pull request

**Comment by AndreiSva** [comment]
what is blocking this pr?

**Review by alois31 — APPROVED** [review]
I can confirm this fixes icons and fonts in Flatpak apps on my system. The changes also look sane to the extent I can say that without familiarity with glib code.

**Comment by beh-10257** [comment]
So this will get merged now ??

**Comment by alois31** [comment]
> So this will get merged now ??

No, this has to be done by someone with commit rights. Please be more patient.

**Review by Henry-Hiles — APPROVED** [review]
LGTM

**Comment by andrevmatos** [comment]
@SuperSandro2000 could you please give a look at this PR? I've been running it for months with great success.

**Review by Ramblurr — APPROVED** [review]
Works for me too!

**Review by SuperSandro2000 — APPROVED** [review]
LGTM but I have no clue about the C code :/

**Comment by beh-10257** [comment]
@andrevmatos can you post just the bit when you overrided the module with this pr
since its been 3 months and there is no guarantee that this will be merged anytime soon I wanna check it out

**Comment by andrevmatos** [comment]
@beh-10257 my system is a flake using https://github.com/gytis-ivaskevicius/flake-utils-plus, and I use the `channels.<name>.patches` functionality of it to apply arbitrary PRs to my system's `nixpkgs` (tracking `nixos-unstable-small`).

**Comment by SuperSandro2000** [comment]
@ofborg build flatpak

**Comment by RAVENz46** [comment]
Fonts fine but icons are still missing.
Is this because I installed icons via home-manager?

**Comment by SadSock** [comment]
How to apply this patch in nixos 23.11?

**Pull request merged**
